oneoffs/prepare_bigquery.py

```diff
@@ -0,0 +1,185 @@
+
+
+The load commands look like:
+bq load --project_id=$PROJECT_ID \
+    --source_format=NEWLINE_DELIMITED_JSON \
+    $PROJECT_ID:minigo_v5_19.games \
+    gs://$BUCKET_NAME/bigquery/holdout/games/* \
+    oneoffs/bigquery_games_schema.json
+
+bq load --project_id=$PROJECT_ID \
+    --source_format=NEWLINE_DELIMITED_JSON \
+    $PROJECT_ID:minigo_v5_19.moves \
+    gs://$BUCKET_NAME/bigquery/holdout/moves/* \
+    oneoffs/bigquery_moves_schema.json
+
+'''
+import collections
+import itertools
+import json
+import os
+import re
+import sys; sys.path.insert(0, '.')
+import sgf
+from tensorflow import gfile
+
+import coords
+import shipname
+import sgf_wrapper
+import utils
+
+DebugRow = collections.namedtuple('DebugRow', [
+    'move', 'action', 'Q', 'U', 'prior', 'orig_prior', 'N', 'soft_N', 'p_delta', 'p_rel'])
+
+BUCKET_NAME = os.environ['BUCKET_NAME']
+GCS_HOLDOUT_PATH = 'gs://%s/data/holdout/' % BUCKET_NAME
+GCS_PATH_TEMPLATE = 'gs://%s/sgf/{}/full/{}' % BUCKET_NAME
+OUTPUT_PATH_TEMPLATE = 'gs://%s/bigquery/holdout/{}/{}' % BUCKET_NAME
+
+def extract_holdout_generation(generation_name):
+    tfrecord_paths = gfile.ListDirectory(GCS_HOLDOUT_PATH + generation_name)
+    filenames = [os.path.basename(path) for path in tfrecord_paths]
+    sgf_names = [GCS_PATH_TEMPLATE.format(generation_name, filename.replace('.tfrecord.zz', '.sgf'))
+        for filename in filenames]
+    game_output_path = OUTPUT_PATH_TEMPLATE.format('games', generation_name)
+    move_output_path = OUTPUT_PATH_TEMPLATE.format('moves', generation_name)
+    with gfile.GFile(game_output_path, 'w') as game_f:
+        with gfile.GFile(move_output_path, 'w') as move_f:
+            for sgf_name in sgf_names:
+                game_data, move_data = extract_data(sgf_name)
+                game_f.write(json.dumps(game_data))
+                game_f.write('\n')
+                for move_datum in move_data:
+                    move_f.write(json.dumps(move_datum))
+                    move_f.write('\n')
+                print('processed {}'.format(sgf_name))
+
+
+def extract_data(filename):
+    with gfile.GFile(filename) as f:
+        contents = f.read()
+        root_node = sgf_wrapper.get_sgf_root_node(contents)
+    game_data = extract_game_data(filename, root_node)
+    move_data = extract_move_data(
+        root_node, game_data['worker_id'], game_data['completed_time'],
+        game_data['board_size'])
+    return game_data, move_data
+
+def extract_game_data(gcs_path, root_node):
+    props = root_node.properties
+    komi = float(sgf_wrapper.sgf_prop(props.get('KM')))
+    result = sgf_wrapper.sgf_prop(props.get('RE', ''))
+    board_size = int(sgf_wrapper.sgf_prop(props.get('SZ')))
+    value = utils.parse_game_result(result)
+    was_resign = '+R' in result
+    
+    filename = os.path.basename(gcs_path)
+    filename_no_ext, _ = os.path.splitext(filename)
+    completion_time = int(filename_no_ext.split('-')[0])
+    worker_id = filename_no_ext.split('-')[-1]
```

Comment:
BTW completion_time & worker_id aren't 100% unique over the self-play games.
I know at least one of the overlaps was a v3 game overlapping with a v5 game

Reply:
Wow. Damn birthday paradox =/

Well worst thing that can happen is that those games are counted 2x in the bulk statistics, but it should be super rare, and so it shouldn't affect them significantly.